text/0000-fmt-size-hint.md

```diff
@@ -0,0 +1,106 @@
+- Start Date: 2015-01-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a `size_hint` method to each of the `fmt` traits, allowing a buffer to allocate with the correct size before writing.
+
+# Motivation
+
+Using the `fmt` traits is slower than a straight `memcpy` of data. The removal of `std::is_utf8` has helped some. The other low-hanging fruit is to add a size hint, so as to reduce or prevent unnecessary allocations during writing to the output buffer. My initial implementation includes [benchmarks][], which looked like this on machine:
+
+[benchmarks]: https://gist.github.com/seanmonstar/8fb7aa6b0512b80522f9#file-size_hint-rs-L91-L162
+
+```
+running 11 tests
+test bench_long         ... bench:       133 ns/iter (+/- 18)
+test bench_long_hint    ... bench:        72 ns/iter (+/- 10)
+test bench_long_memcpy  ... bench:        44 ns/iter (+/- 2)
+test bench_med          ... bench:       112 ns/iter (+/- 10)
+test bench_med_hint     ... bench:        59 ns/iter (+/- 7)
+test bench_med_memcpy   ... bench:        32 ns/iter (+/- 6)
+test bench_nested       ... bench:       248 ns/iter (+/- 19)
+test bench_nested_hint  ... bench:       134 ns/iter (+/- 6)
+test bench_short        ... bench:        96 ns/iter (+/- 13)
+test bench_short_hint   ... bench:        60 ns/iter (+/- 3)
+test bench_short_memcpy ... bench:        33 ns/iter (+/- 3)
+```
+
+# Detailed design
```

Comment:
I think this section is currently missing details about how this is actually going to be implemented in relation to `format_args!` unless we're *only* optimizing the `.to_string()` case. For example, these two calls are fairly different in the paths that they take:

```rust
foo.to_string();
format!("{}", foo);
```

In the first case, we call `to_string` on the `ToString` trait which has a reference to the concrete type `T`, allowing an invocation of `size_hint` to preallocate the `String`. In the latter case we call `std::fmt::format` with an `Arguments` structure. In `Arguments` the type `T` has been erased, and there is not an obvious location to call `.size_hint()`.

I would be curious to see more details about how this is handled and how far these size hints are being propagated.